contracts/adapters/swap/lido-satellite/src/contract.rs

```diff
@@ -0,0 +1,121 @@
+};
+use cosmwasm_std::{entry_point, to_binary, DepsMut, Env, MessageInfo, Response, WasmMsg};
+use cw_utils::one_coin;
+use skip::swap::{
+    execute_transfer_funds_back, ExecuteMsg, LidoSatelliteInstantiateMsg as InstantiateMsg,
+    SwapOperation,
+};
+
+///////////////////
+/// INSTANTIATE ///
+///////////////////
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> ContractResult<Response> {
+    // Validate entry point contract address
+    let checked_entry_point_contract_address =
+        deps.api.addr_validate(&msg.entry_point_contract_address)?;
+
+    // Store the entry point contract address
+    ENTRY_POINT_CONTRACT_ADDRESS.save(deps.storage, &checked_entry_point_contract_address)?;
+
+    // Validate satellite contract address
+    let checked_lido_satellite_contract_address = deps
+        .api
+        .addr_validate(&msg.lido_satellite_contract_address)?;
+
+    // Store the satellite contract address
+    LIDO_SATELLITE_CONTRACT_ADDRESS.save(deps.storage, &checked_lido_satellite_contract_address)?;
+
+    Ok(Response::new()
+        .add_attribute("action", "instantiate")
+        .add_attribute(
+            "entry_point_contract_address",
+            checked_entry_point_contract_address.to_string(),
+        )
+        .add_attribute(
+            "lido_satellite_contract_address",
+            checked_lido_satellite_contract_address.to_string(),
+        ))
+}
+
+///////////////
+/// EXECUTE ///
+///////////////
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> ContractResult<Response> {
+    match msg {
+        ExecuteMsg::Swap { operations } => execute_swap(deps, env, info, operations),
+        ExecuteMsg::TransferFundsBack { swapper } => {
+            Ok(execute_transfer_funds_back(deps, env, info, swapper)?)
+        }
+    }
+}
+
+fn execute_swap(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    // FIXME: seems like we are doomed to ignore this field at all?
+    _operations: Vec<SwapOperation>,
+) -> ContractResult<Response> {
+    // Get entry point contract address from storage
+    let entry_point_contract_address = ENTRY_POINT_CONTRACT_ADDRESS.load(deps.storage)?;
+
+    // Enforce the caller is the entry point contract
+    if info.sender != entry_point_contract_address {
+        return Err(ContractError::Unauthorized);
+    }
+
+    // Get coin in from the message info, error if there is not exactly one coin sent
+    let coin_in = one_coin(&info)?;
+
+    let lido_satellite_contract_address = LIDO_SATELLITE_CONTRACT_ADDRESS.load(deps.storage)?;
+    let lido_satellite_config: lido_satellite::msg::ConfigResponse =
+        deps.querier.query_wasm_smart(
+            &lido_satellite_contract_address,
+            &lido_satellite::msg::QueryMsg::Config {},
+        )?;
```

Comment:
If this is immutable on the lido side, do you think it makes sense to add the bridge_denom and canonical_denom in the instantiation method instead, saving us an external contract query per call? (as well as making testing a tad easier)

If there is any expectation this is not immutable, I'd err on the side of keeping this as is to avoid any future issues

Reply:
I've actually been thinking about this one! Yes, I like the idea of caching these values during the instantiation.

While canonical denom will never change, we plan to abolish bridged denom eventually, so this adapter will stop working anyway no matter if we cache denoms or not. So, yeah, let's just cache them. I will implement it shortly.